squad/api/rest.py

```diff
@@ -1310,9 +1310,17 @@ class StatusViewSet(NestedViewSetMixin, ModelViewSet):
 
 class AttachmentSerializer(serializers.ModelSerializer):
 
+    download_url = serializers.SerializerMethodField()
+
     class Meta:
         model = Attachment
-        fields = ('filename', 'mimetype', 'length')
+        fields = ('download_url', 'filename', 'mimetype', 'length')
+
+    def get_download_url(self, attachment):
+        request = self.context.get('request')
+        if request is None:
+            return None
+        return request.build_absolute_uri(f"/{attachment.test_run.project}/build/{attachment.test_run.build.version}/attachments/testrun/{attachment.test_run.pk}/{attachment.filename}")
```

Comment:
I thought about that as well, but the problem with using file name as part of the url could be troublesome as it can contain funny characters.

The way I implemented, one can see all attachments for a testrun using `/api/testruns/123/`, then download individual attachments by constructing the URL `/api/testruns/123/attachments?filename=the-file-name`. In this case, the file name can contain whatever characters and they'll all be url-encoded.

Did you want links in each attachment so you wouldn't have to generate urls manually?

Reply:
Yes - I didn't actually realise it was possible to download the attachments this way (if you can point me towards the code that makes this possible that would be great).

I think it would be easiest to work with if there were links directly to the download URLs of the files, as it seems quite unintuitive/error-prone to have to manually construct the URL.

Regarding the issue with funny characters - wouldn't this be an issue regardless of whether we have a download URL stored in SQUAD? (aka if someone puts a space in a filename, it is going to be a problem when doing the "...?filename=..." method, too?)

Reply:
I think you could do that with something like this:
```python
import urllib.parse

...

    def get_download_url(self, attachment):
        request = self.context.get('request')
        if request is None:
            return None
        base_url = rest_reverse('testrun-detail', pk=attachment.test_run.pk, request=request)
        filename_url_encoded = urllib.parse.quote(attachment.filename, safe='')
        return f'{base_url}/attachments?filename=filename_url_encded'
```

How does it look?

Reply:
Looks good to me - looks similar to another approach I put together, actually (but I was just missing the fact the attachments could be downloaded with "?filename=...")

This was my other approach if you were interested:
```
    def get_download_url(self, attachment):
        request = self.context.get('request')
        group = str(attachment.test_run.project).split('/')[0]
        project = str(attachment.test_run.project).split('/')[1]
        args = (group, 
                  project, 
                  attachment.test_run.build.version, 
                  attachment.test_run.pk, 
                  attachment.filename)
        relative_url = reverse('build_attachments', args=args)
        return request.build_absolute_uri(relative_url)
```

I will give your suggested method a try :)

I guess this part of the code can also be removed:
```
        request = self.context.get('request')
        if request is None:
            return None
```
I didn't realised at first that using `rest_reverse` would remove the need for the `self.context.get('request')`, but it seems obvious now :)

Reply:
actually just spotted that request is needed for rest_reverse, it's just that in the other places where it is used, request is passed in as an arg to the function - had just been wondering why the `request = self.context.get('request')` line was not used elsewhere in the file.